### PR TITLE
ARROW-4124: [C++] Draft Aggregate and Sum kernels

### PR DESCRIPTION
**Pull request by fsaintjacques**
This is a draft of the aggregate kernel interface. The goal of publishing this PR is to gather feedback on the design and architecture.

- `AggregateUnaryKernel/AggregateState` decomposition kernel implementation and parallel execution. The implementor of an `AggregateState` does not need to be concerned about details of parallel execution (minus the thread-safety of the `Consume` method). It also allows user to implement custom aggregates. 
- `Monoid` a class representing a mathematical monoid.
- `MonoidAggregateState` a specific implementation of `AggregateState` that is generic enough to support various monoids. Note that it is limited to monoid known at compile time, usually monoids on primitives.

Currently only the "Sum" kernel is implemented. Further kernels will be implemented in follow up patches.

The current draft does not take into account filtering and groups. I do have an idea of how this would be done, e.g. add `AggregateState::Consume(..., const Mask &filter)`. I delayed incorporating this interface until we implement the filtering kernel. This is going to be done in a different ticket.

### CODE REVIEW

File: cpp/src/arrow/compute/kernel.h
Comment (bkietz) [inline cpp/src/arrow/compute/kernel.h]: Couldn't you just `static_cast<Type::type>(this->value.which())`?

File: cpp/src/arrow/compute/kernels/monoid-impl.h
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/monoid-impl.h]: These should be undefed below

File: cpp/src/arrow/compute/kernels/aggregation.cc
Comment (emkornfield) [inline cpp/src/arrow/compute/kernels/aggregation.cc]: this seems like a slightly different pattern then was taken for UnaryArrayKernel (i.e. this pattern wasn't in the kernel, but instead in util-internal.h).  I haven't thought through if a similar pattern should try to be established, have you?

File: cpp/src/arrow/compute/kernels/aggregation.h
Comment (emkornfield) [inline cpp/src/arrow/compute/kernels/aggregation.h]: I think this is only true for Monoids?

File: cpp/src/arrow/compute/kernels/monoid-impl.h
Comment (emkornfield) [inline cpp/src/arrow/compute/kernels/monoid-impl.h]: Are compilers generally good enough to optimize out any overhead from calling the constructor (I would assume so), or would it make sense to have have a slightly less clean Monoid abstraction that takes the raw ValueType?

File: cpp/src/arrow/compute/kernels/monoid-impl.h
Comment (emkornfield) [inline cpp/src/arrow/compute/kernels/monoid-impl.h]: I'm not sure if the discussion happened already but it would be good to understand the general plan around threading model, and how this relates the physical execution layer (i.e. something like the volcano model) and what exactly should live in which level (i.e. what belongs in the kernel/state).

I'm sure you considered this, but a different design might create a kernel per thread, then merge each scalar at the end if the work can in fact can be run in parallel.  Probably not much difference performance wise but it still might eliminate an unnecessary memory barrier/system call.

File: cpp/src/arrow/compute/kernels/monoid-impl.h
Comment (emkornfield) [inline cpp/src/arrow/compute/kernels/monoid-impl.h]: Do monoids always have to be a scalar?  Would it make sense to have monoid_ populate the datum instead of assuming this?

File: cpp/src/arrow/compute/kernels/sum-test.cc
Comment (emkornfield) [inline cpp/src/arrow/compute/kernels/sum-test.cc]: nit: CheckSum is a little bit confusing name due to checksum being  a potential operation.  Something like VerifySum might read a little better.

**Review by emkornfield — COMMENTED** [review]
Not an expert in this space so take my comments with a grain of salt.

**Comment by wesm** [comment]
I will take a closer look, but one comment: thread control / multithreading should be handled (IMHO) at a higher level than the aggregation kernels.

File: cpp/src/arrow/compute/kernels/aggregation.cc
Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/aggregation.cc]: I didn't see `util-internal.h`, I'll try to merge both. When I used this piece, I would have preferred to not be concerned with this and hoped that `Datum` provided a ChunkedArray for both cases (wrapping in a single element ChunkedArray when datum is Array).

I'll add this feature.

File: cpp/src/arrow/compute/kernel.h
Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernel.h]: We'd have to make sure the mapping is exact, which is almost equivalent to expliciting this list.

File: cpp/src/arrow/compute/kernels/monoid-impl.h
Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/monoid-impl.h]: The compiler is clever enough (at least with O3). But, we can also add a second `operator+=` that takes the scalar value .

Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/monoid-impl.h]: Also note that some Monoid will be composed and aren't made of a single scalar value, e.g. the `Mean` monoid is made of `std::tuple<size_t, T>`.

Comment (emkornfield) [reply cpp/src/arrow/compute/kernels/monoid-impl.h]: 👍 cool, its nice to see the compiler can auto-vectorize this.

File: cpp/src/arrow/compute/kernels/sum-benchmark.cc
Comment (emkornfield) [inline cpp/src/arrow/compute/kernels/sum-benchmark.cc]: might be nice to have a benchmark that includes some nulls

File: cpp/src/arrow/compute/kernels/sum-test.cc
Comment (emkornfield) [inline cpp/src/arrow/compute/kernels/sum-test.cc]: one interesting case is all nulls.

File: cpp/src/arrow/compute/kernels/sum-benchmark.cc
Comment (wesm) [reply cpp/src/arrow/compute/kernels/sum-benchmark.cc]: See some benchmarking I did about this http://wesmckinney.com/blog/bitmaps-vs-sentinel-values/

Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/sum-benchmark.cc]: @emkornfield I will add both, my first benchmark was with nulls, but I removed them to to benchmark with perf.

File: cpp/src/arrow/compute/kernels/sum-test.cc
Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/sum-test.cc]: Yes, and it's already an issue with the current interface, e.g. the min/max monoid would return `std::numeric_limits<T>::max/min` respectively and that is an issue; it should return NULL.

File: cpp/src/arrow/compute/kernels/aggregation.h
Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/aggregation.h]: Do you have an example of an aggregate function which is not associative?

Comment (emkornfield) [reply cpp/src/arrow/compute/kernels/aggregation.h]: Short answer (might be contrived): "exponentially smoothed moving average" and "hash elements"
Longer answer (probably obvious and too pedantic):   I was assuming a naive parallelization where Consume could be called in any order which would require Commutativity and associativity. With non-commutative associative operations (e.g. string concatenation, FIRST, LAST, etc ) there are more book-keeping details to to ensure parallelization still returns the correct results.  At the moment I'm having a hard time coming up with an example that benefits from parallelization of onsume that is not commutative.    

From a practical standpoint for this review:
*It might be good is add some additional pseudo-code for the parallelism AggregateState is meant to handle, * Possibly having some mechanism to signal commutative and non-commutative state.

File: cpp/src/arrow/compute/kernel.h
Comment (emkornfield) [inline cpp/src/arrow/compute/kernel.h]: Would it make sense to include string and unicode types in here?

**Comment by wesm** [comment]
Will review soon

File: cpp/src/arrow/compute/kernel.h
Comment (emkornfield) [reply cpp/src/arrow/compute/kernel.h]: Filed ARROW-4331 to track adding additional types.

File: cpp/src/arrow/compute/kernels/monoid-impl.h
Comment (emkornfield) [reply cpp/src/arrow/compute/kernels/monoid-impl.h]: Created ARROW-4333 to track this.

**Comment by fsaintjacques** [comment]
As @emkornfield proposed, let's move threading/architecture of execution in https://issues.apache.org/jira/browse/ARROW-4333.

**Review by wesm — CHANGES_REQUESTED** [review]
Unfortunately, this implementation has branching in the inner loop, so the approach here might have to be rethought. The performance difference by removing branching and partially unrolling the inner loop is _huge_. See http://wesmckinney.com/blog/bitmaps-vs-sentinel-values/

**Comment by wesm** [comment]
Another question that needs to be resolved is how to return a NULL value. What we do in pandas is have a "min_values" option, so:

* sum of all nulls is 0 by default, but you can set `min_values=1` if you want it to be NULL instead
* mean of all nulls is NULL (`min_values=1` by default)

File: cpp/src/arrow/compute/kernels/aggregation.h
Comment (emkornfield) [reply cpp/src/arrow/compute/kernels/aggregation.h]: To answer my own question about associate and not commutative multithreading, the parsing of CSV files is a good example where multithreading makes sense.

File: cpp/src/arrow/compute/kernels/sum-benchmark.cc
Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/sum-benchmark.cc]: Will do.

File: cpp/src/arrow/compute/kernels/monoid-impl.h
Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/monoid-impl.h]: No since they're "exported" macros in other files.

File: cpp/src/arrow/compute/kernels/aggregation.cc
Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/aggregation.cc]: Addressed.

**Comment by wesm** [comment]
@fsaintjacques the PR description is out of date, can you update? I'm going to review and leave comments so we can merge this and move on to the next steps

File: cpp/src/arrow/util/bit-util.h
Comment (wesm) [inline cpp/src/arrow/util/bit-util.h]: Add unit tests for this

**Review by wesm — COMMENTED** [review]
Some minor comments for follow up patches. I'm going to tweak a few things and then merge this once the build is passing

File: cpp/src/arrow/compute/kernels/sum-benchmark.cc
Comment (wesm) [reply cpp/src/arrow/compute/kernels/sum-benchmark.cc]: I just did this

File: cpp/src/arrow/compute/kernels/CMakeLists.txt
Comment (wesm) [reply cpp/src/arrow/compute/kernels/CMakeLists.txt]: done

File: cpp/src/arrow/compute/kernels/aggregate.h
Comment (wesm) [reply cpp/src/arrow/compute/kernels/aggregate.h]: done

File: cpp/src/arrow/compute/kernels/sum-benchmark.cc
Comment (wesm) [reply cpp/src/arrow/compute/kernels/sum-benchmark.cc]: done

File: cpp/src/arrow/compute/kernels/sum-benchmark.cc
Comment (wesm) [reply cpp/src/arrow/compute/kernels/sum-benchmark.cc]: done

File: cpp/src/arrow/compute/kernels/sum.cc
Comment (wesm) [reply cpp/src/arrow/compute/kernels/sum.cc]: done

**Review by wesm — APPROVED** [review]

**Comment by wesm** [comment]
@fsaintjacques I tweaked the PR description. I'll merge this once the build passes in the interest of project velocity. There's a lot of follow up issues to create so I'll leave that to you

**Comment by wesm** [comment]
The benchmark results on my machine for posterity:

```
---------------------------------------------------------------------------------------------------------------
Benchmark                                                        Time           CPU Iterations UserCounters...
---------------------------------------------------------------------------------------------------------------
BenchSum<SumNoNullsUnrolled<int64_t>>/32768/0                    1 us          1 us     833069 null_percent=0 size=32.768k   37.2508GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/32768/1                    1 us          1 us     838507 null_percent=1 size=32.768k   37.5707GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/32768/10                   1 us          1 us     833487 null_percent=10 size=32.768k   37.6234GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/32768/50                   1 us          1 us     691232 null_percent=50 size=32.768k   36.0022GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/262144/0                  12 us         12 us      57468 null_percent=0 size=262.144k   20.2227GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/262144/1                  12 us         12 us      57788 null_percent=1 size=262.144k     19.83GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/262144/10                 12 us         12 us      58093 null_percent=10 size=262.144k   20.1576GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/262144/50                 12 us         12 us      56866 null_percent=50 size=262.144k   20.0273GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/8388608/0                509 us        509 us       1421 null_percent=0 size=8.38861M   15.3606GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/8388608/1                504 us        504 us       1360 null_percent=1 size=8.38861M   15.5096GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/8388608/10               489 us        489 us       1425 null_percent=10 size=8.38861M   15.9891GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/8388608/50               517 us        517 us       1000 null_percent=50 size=8.38861M   15.1099GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/33554432/0              2336 us       2336 us        275 null_percent=0 size=33.5544M   13.3768GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/33554432/1              2344 us       2343 us        300 null_percent=1 size=33.5544M   13.3348GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/33554432/10             2337 us       2337 us        297 null_percent=10 size=33.5544M   13.3746GB/s
BenchSum<SumNoNullsUnrolled<int64_t>>/33554432/50             2404 us       2404 us        299 null_percent=50 size=33.5544M   13.0007GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/32768/0                   3 us          3 us     218287 null_percent=0 size=32.768k   9.51592GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/32768/1                   3 us          3 us     216732 null_percent=1 size=32.768k    9.4701GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/32768/10                  3 us          3 us     218117 null_percent=10 size=32.768k   9.52378GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/32768/50                  3 us          3 us     216833 null_percent=50 size=32.768k   9.45774GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/262144/0                 26 us         26 us      27222 null_percent=0 size=262.144k   9.53919GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/262144/1                 26 us         26 us      27233 null_percent=1 size=262.144k   9.52018GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/262144/10                26 us         26 us      27106 null_percent=10 size=262.144k   9.52412GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/262144/50                26 us         26 us      27039 null_percent=50 size=262.144k   9.52342GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/8388608/0               866 us        866 us        802 null_percent=0 size=8.38861M   9.02228GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/8388608/1               865 us        865 us        792 null_percent=1 size=8.38861M   9.03213GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/8388608/10              865 us        865 us        808 null_percent=10 size=8.38861M   9.02746GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/8388608/50              866 us        866 us        807 null_percent=50 size=8.38861M   9.02335GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/33554432/0             3541 us       3541 us        198 null_percent=0 size=33.5544M   8.82504GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/33554432/1             3541 us       3540 us        197 null_percent=1 size=33.5544M   8.82724GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/33554432/10            3545 us       3545 us        198 null_percent=10 size=33.5544M   8.81579GB/s
BenchSum<SumSentinelUnrolled<int64_t>>/33554432/50            3543 us       3543 us        198 null_percent=50 size=33.5544M   8.81922GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/32768/0              1 us          1 us     530767 null_percent=0 size=32.768k   23.1373GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/32768/1              1 us          1 us     480515 null_percent=1 size=32.768k   20.9358GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/32768/10             2 us          2 us     328397 null_percent=10 size=32.768k   14.2905GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/32768/50             3 us          3 us     255490 null_percent=50 size=32.768k   11.1082GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/262144/0            14 us         14 us      50120 null_percent=0 size=262.144k   17.4662GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/262144/1            15 us         15 us      47530 null_percent=1 size=262.144k   16.6678GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/262144/10           21 us         21 us      33159 null_percent=10 size=262.144k    11.662GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/262144/50           23 us         23 us      30148 null_percent=50 size=262.144k   10.5367GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/8388608/0          553 us        553 us       1149 null_percent=0 size=8.38861M   14.1303GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/8388608/1          638 us        638 us       1101 null_percent=1 size=8.38861M   12.2365GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/8388608/10        1148 us       1148 us        612 null_percent=10 size=8.38861M   6.80722GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/8388608/50         780 us        780 us        893 null_percent=50 size=8.38861M   10.0146GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/33554432/0        2513 us       2513 us        281 null_percent=0 size=33.5544M   12.4369GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/33554432/1        2854 us       2854 us        246 null_percent=1 size=33.5544M   10.9505GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/33554432/10       5169 us       5170 us        100 null_percent=10 size=33.5544M   6.04495GB/s
BenchSum<SumBitmapVectorizeUnroll<int64_t>>/33554432/50       3231 us       3231 us        207 null_percent=50 size=33.5544M   9.67302GB/s
```

The L1-cache-sized benchmarks seem like they can be a bit noisy

File: cpp/src/arrow/compute/kernels/aggregate.h
Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/aggregate.h]: The intent was to have a strongly typed parameters and let the error handling on the caller.

File: cpp/src/arrow/compute/kernels/CMakeLists.txt
Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/CMakeLists.txt]: :+1:, such that we amortize the slow linking issue with windows?

File: cpp/src/arrow/compute/kernels/aggregate.h
Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/aggregate.h]: Because you do not wanted shared-ptr boxed primitives for group-by results, you write into a Buffer (which can then be converted to a proper typed Array). This is critical for cases where the group expression does not reduce the data size.

The second issue is that some operators requires extra allocation for each results (`think HyperLogLog(precision)`, or `Distribution(max_vals))`. This poses 2 sub-problems, the size of the state may require extra allocation (and thus reclamation via destructor), and the size might be a function of a non-static argument.

The following query would be untenable with `shared_ptr<State>`:

```SQL
SELECT
  campaign_id,
  user_id,
  SUM(impressions),
  SUM(clicks),
  HYPERLOGLOG(sites, bins = 64, precision_bits=8), -- size is 64 bytes
  DISTRIBUTION(sites, top_bins = 50) -- size is struct<String, size_t>[50]
GROUP BY   -- Very high cardinality groups
  campaign_id,
  user_id; 
```

File: cpp/src/arrow/compute/kernels/CMakeLists.txt
Comment (wesm) [reply cpp/src/arrow/compute/kernels/CMakeLists.txt]: among other things, yeah. I don't think we should have one test executable for each kernel, but rather to organize tests according to logical relationships (e.g. "mean" is close to "sum")

File: cpp/src/arrow/compute/kernels/aggregate.h
Comment (wesm) [reply cpp/src/arrow/compute/kernels/aggregate.h]: I still don't understand the need for the elaborate indirection. I'm sorry to be a bit thick, let's address it once we implement some more complex aggregation functions

File: cpp/src/arrow/compute/kernels/aggregate-benchmark.cc
Comment (fsaintjacques) [inline cpp/src/arrow/compute/kernels/aggregate-benchmark.cc]: The previous version was faster, clang-6 and onward auto-vectorize this perfectly (even with just O3): https://godbolt.org/z/v5nFNM . The result would be even better with AVX512 enabled (see https://godbolt.org/z/LFcQJv and  https://llvm.org/devmtg/2015-04/slides/MaskedIntrinsics.pdf).

**Pull request closed without merging**
